// angular-inline-editors/dist/date-editor/date-editor.module.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
const core_1 = require("@angular/core");
const forms_1 = require("@angular/forms");
const common_1 = require("@angular/common");
const date_editor_component_1 = require("./date-editor.component");
const ngx_bootstrap_1 = require("ngx-bootstrap");
let DateEditorModule = DateEditorModule_1 = class DateEditorModule {
    static forRoot() {
        return {
            ngModule: DateEditorModule_1,
        };
    }
};
DateEditorModule = DateEditorModule_1 = __decorate([
    core_1.NgModule({
        imports: [
            common_1.CommonModule,
            forms_1.FormsModule,
            ngx_bootstrap_1.BsDatepickerModule.forRoot()
        ],
        declarations: [
            date_editor_component_1.DateEditorComponent
        ],
        exports: [date_editor_component_1.DateEditorComponent]
    })
], DateEditorModule);
exports.DateEditorModule = DateEditorModule;
var DateEditorModule_1;
