// angular-inline-editors/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var input_editor_module_1 = require("./input-editor/input-editor.module");
exports.InputEditorModule = input_editor_module_1.InputEditorModule;
var textarea_editor_module_1 = require("./textarea-editor/textarea-editor.module");
exports.TextAreaEditorModule = textarea_editor_module_1.TextAreaEditorModule;
var select_editor_module_1 = require("./select-editor/select-editor.module");
exports.SelectEditorModule = select_editor_module_1.SelectEditorModule;
var checklist_editor_module_1 = require("./checklist-editor/checklist-editor.module");
exports.CheckListEditorModule = checklist_editor_module_1.CheckListEditorModule;
var radiolist_editor_module_1 = require("./radiolist-editor/radiolist-editor.module");
exports.RadioListEditorModule = radiolist_editor_module_1.RadioListEditorModule;
var checkbox_editor_module_1 = require("./checkbox-editor/checkbox-editor.module");
exports.CheckBoxEditorModule = checkbox_editor_module_1.CheckBoxEditorModule;
var tags_editor_module_1 = require("./tags-editor/tags-editor.module");
exports.TagsEditorModule = tags_editor_module_1.TagsEditorModule;
var date_editor_module_1 = require("./date-editor/date-editor.module");
exports.DateEditorModule = date_editor_module_1.DateEditorModule;
var time_editor_module_1 = require("./time-editor/time-editor.module");
exports.TimeEditorModule = time_editor_module_1.TimeEditorModule;
var typeahead_editor_module_1 = require("./typeahead-editor/typeahead-editor.module");
exports.TypeAheadEditorModule = typeahead_editor_module_1.TypeAheadEditorModule;
var number_editor_module_1 = require("./inputnumber-editor/number-editor.module");
exports.NumberEditorModule = number_editor_module_1.NumberEditorModule;
var datetimepicker_editor_module_1 = require("./datetimepicker-editor/datetimepicker-editor.module");
exports.DateTimeEditorModule = datetimepicker_editor_module_1.DateTimeEditorModule;
