// ng2-file-input/dist/bundles/ng2-file-input.umd.js
(function (global, factory) {
	typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@angular/platform-browser'), require('@angular/common'), require('bergben-angular2-file-drop')) :
	typeof define === 'function' && define.amd ? define(['exports', '@angular/core', '@angular/platform-browser', '@angular/common', 'bergben-angular2-file-drop'], factory) :
	(factory((global['ng2-file-input'] = {}),global.ng.core,global.platformBrowser,global.common,global.bergbenAngular2FileDrop));
}(this, (function (exports,core,platformBrowser,common,bergbenAngular2FileDrop) { 'use strict';

var FileInputHandlerService = (function () {
    function FileInputHandlerService() {
        this.fileInputs = [];
    }
    FileInputHandlerService.prototype.add = function (id) {
        this.fileInputs.push({
            id: id,
            currentFiles: []
        });
        return true;
    };
    FileInputHandlerService.prototype.remove = function (id) {
        for (var i = 0; this.fileInputs.length; i++) {
            if (this.fileInputs[i].id === id) {
                this.fileInputs.splice(i, 1);
                i = this.fileInputs.length;
                return true;
            }
        }
        return false;
    };
    FileInputHandlerService.prototype.reset = function (id) {
        var fileInput = this.getFileInput(id);
        fileInput.currentFiles = [];
        return true;
    };
    FileInputHandlerService.prototype.addFiles = function (fileInputId, files) {
        var _this = this;
        var fileInput = this.getFileInput(fileInputId);
        var notAdded = [];
        files.forEach(function (file) {
            if (!_this.addFile(file, fileInput.currentFiles)) {
                notAdded.push(file);
            }
        });
        return notAdded;
    };
    FileInputHandlerService.prototype.removeFiles = function (fileInputId, files) {
        var _this = this;
        var fileInput = this.getFileInput(fileInputId);
        var notRemoved = [];
        files.forEach(function (file) {
            if (!_this.removeFile(file, fileInput.currentFiles)) {
                notRemoved.push(file);
            }
        });
        return notRemoved;
    };
    FileInputHandlerService.prototype.getFileInput = function (id) {
        for (var i = 0; this.fileInputs.length; i++) {
            if (this.fileInputs[i].id === id) {
                return this.fileInputs[i];
            }
        }
    };
    FileInputHandlerService.prototype.addFile = function (file, toFiles) {
        toFiles.push(file);
        return true;
    };
    FileInputHandlerService.prototype.removeFile = function (file, fromFiles) {
        for (var i = 0; fromFiles.length; i++) {
            if (this.isSameFile(file, fromFiles[i])) {
                fromFiles.splice(i, 1);
                i = fromFiles.length;
                return true;
            }
        }
        return false;
    };
    FileInputHandlerService.prototype.isSameFile = function (file1, file2) {
        return file1.name === file2.name && file1.size === file2.size && file1.type === file2.type;
    };
    FileInputHandlerService.decorators = [
        { type: core.Injectable },
    ];
    /** @nocollapse */
    FileInputHandlerService.ctorParameters = function () { return []; };
    return FileInputHandlerService;
}());

var Ng2FileInputOptions = (function () {
    function Ng2FileInputOptions() {
        this.dropText = "Drop file here";
        this.browseText = "Browse";
        this.removeText = "Remove";
        this.invalidFileText = "You have picked an invalid or disallowed file.";
        this.invalidFileTimeout = 8000;
        this.removable = true;
        this.multiple = false;
        this.accept = '';
        this.showPreviews = true;
        this.extensions = [];
    }
    return Ng2FileInputOptions;
}());

(function (Ng2FileInputAction) {
    Ng2FileInputAction[Ng2FileInputAction["Removed"] = 0] = "Removed";
    Ng2FileInputAction[Ng2FileInputAction["Added"] = 1] = "Added";
    Ng2FileInputAction[Ng2FileInputAction["InvalidDenied"] = 2] = "InvalidDenied";
    Ng2FileInputAction[Ng2FileInputAction["CouldNotRemove"] = 3] = "CouldNotRemove";
    Ng2FileInputAction[Ng2FileInputAction["CouldNotAdd"] = 4] = "CouldNotAdd";
})(exports.Ng2FileInputAction || (exports.Ng2FileInputAction = {}));

var Ng2FileInputComponent = (function () {
    function Ng2FileInputComponent(fileInputHandlerService, sanitizer, defaultOptions) {
        this.fileInputHandlerService = fileInputHandlerService;
        this.sanitizer = sanitizer;
        this.defaultOptions = defaultOptions;
        this.alreadyEmitted = false;
        this.fileIsOver = false;
        this.invalidFile = false;
        this.multiple = null;
        this.removable = null;
        this.showPreviews = null;
        this.outputAction = new core.EventEmitter();
        this.outputRemoved = new core.EventEmitter();
        this.outputAdded = new core.EventEmitter();
        this.outputInvalidDenied = new core.EventEmitter();
        this.outputCouldNotRemove = new core.EventEmitter();
        this.outputCouldNotAdd = new core.EventEmitter();
    }
    Ng2FileInputComponent.prototype.ngOnInit = function () {
        this.dropText = this.dropText || this.defaultOptions.dropText;
        this.browseText = this.browseText || this.defaultOptions.browseText;
        this.removeText = this.removeText || this.defaultOptions.removeText;
        this.invalidFileText = this.invalidFileText || this.defaultOptions.invalidFileText;
        this.invalidFileTimeout = this.invalidFileTimeout || this.defaultOptions.invalidFileTimeout;
        this.multiple = this.multiple !== null ? this.multiple : this.defaultOptions.multiple;
        this.accept = this.accept !== null ? this.accept : this.defaultOptions.accept;
        this.removable = this.removable !== null ? this.removable : this.defaultOptions.removable;
        this.showPreviews = this.showPreviews !== null ? this.showPreviews : this.defaultOptions.showPreviews;
        this.extensions = this.extensions || this.defaultOptions.extensions;
        if (typeof (this.id) === "undefined" || !this.id || this.id === null) {
            this.id = this.generateId();
        }
        this.fileInputHandlerService.add(this.id);
    };
    Ng2FileInputComponent.prototype.ngOnDestroy = function () {
        this.fileInputHandlerService.remove(this.id);
    };
    Ng2FileInputComponent.prototype.fileOver = function (fileIsOver) {
        this.fileIsOver = fileIsOver;
    };
    Ng2FileInputComponent.prototype.onFileDrop = function (file) {
        var _this = this;
        if (!this.multiple) {
            //make sure only to emit once
            if (!this.alreadyEmitted) {
                this.alreadyEmitted = true;
                this.handleFile(file);
                setTimeout(function () { _this.alreadyEmitted = false; }, 0);
            }
        }
        else {
            this.handleFile(file);
        }
    };
    Ng2FileInputComponent.prototype.fileSelected = function (event) {
        this.invalidFile = false;
        var files = event.target.files;
        if (files && files.length) {
            if (!this.multiple) {
                //make sure only to emit one
                this.handleFile(files[0]);
            }
            else {
                for (var i = 0; i < files.length; i++) {
                    this.handleFile(files[i]);
                }
            }
        }
        setTimeout(function () { event.target.value = ""; }, 0);
    };
    Ng2FileInputComponent.prototype.getCurrentFiles = function () {
        var fileInput = this.fileInputHandlerService.getFileInput(this.id);
        return fileInput ? fileInput.currentFiles : [];
    };
    Ng2FileInputComponent.prototype.removeFile = function (file) {
        if (this.removable) {
            var notRemovedFiles = this.fileInputHandlerService.removeFiles(this.id, [file]);
            if (notRemovedFiles.length === 0) {
                this.emitOutput(file, exports.Ng2FileInputAction.Removed);
            }
            else {
                this.emitOutput(file, exports.Ng2FileInputAction.CouldNotRemove);
            }
        }
    };
    Ng2FileInputComponent.prototype.getObjectUrl = function (file) {
        return this.sanitizer.bypassSecurityTrustResourceUrl(window.URL.createObjectURL(file));
    };
    Ng2FileInputComponent.prototype.handleFile = function (file) {
        if (this.isValidFile(file)) {
            if (!this.multiple) {
                this.fileInputHandlerService.reset(this.id);
            }
            var notAddedFiles = this.fileInputHandlerService.addFiles(this.id, [file]);
            if (notAddedFiles.length === 0) {
                this.emitOutput(file, exports.Ng2FileInputAction.Added);
            }
            else {
                this.emitOutput(file, exports.Ng2FileInputAction.CouldNotAdd);
            }
        }
        else {
            this.emitOutput(file, exports.Ng2FileInputAction.InvalidDenied);
        }
    };
    Ng2FileInputComponent.prototype.emitOutput = function (file, action) {
        this.outputAction.emit({
            id: this.id,
            currentFiles: this.getCurrentFiles(),
            action: action,
            file: file
        });
        switch (action) {
            case exports.Ng2FileInputAction.Added:
                this.outputAdded.emit({
                    id: this.id,
                    currentFiles: this.getCurrentFiles(),
                    file: file
                });
                break;
            case exports.Ng2FileInputAction.Removed:
                this.outputRemoved.emit({
                    id: this.id,
                    currentFiles: this.getCurrentFiles(),
                    file: file
                });
                break;
            case exports.Ng2FileInputAction.InvalidDenied:
                this.outputInvalidDenied.emit({
                    id: this.id,
                    currentFiles: this.getCurrentFiles(),
                    file: file
                });
                break;
            case exports.Ng2FileInputAction.CouldNotAdd:
                this.outputCouldNotAdd.emit({
                    id: this.id,
                    currentFiles: this.getCurrentFiles(),
                    file: file
                });
                break;
            case exports.Ng2FileInputAction.CouldNotRemove:
                this.outputCouldNotRemove.emit({
                    id: this.id,
                    currentFiles: this.getCurrentFiles(),
                    file: file
                });
                break;
        }
    };
    Ng2FileInputComponent.prototype.isValidFile = function (file) {
        var _this = this;
        if (this.extensions.length > 0) {
            var ext = file.name.split('.').pop();
            if ((this.extensions.indexOf('image/jpg') !== -1) && (this.extensions.indexOf('image/jpeg') === -1)) {
                this.extensions.push('image/jpeg');
            }
            
            if (this.extensions.indexOf(file.type) === -1 && this.extensions.indexOf(ext) === -1) {
                this.invalidFile = true;
                if (this.invalidFileTimeout !== 0) {
                    setTimeout(function () {
                        _this.invalidFile = false;
                    }, this.invalidFileTimeout);
                }
                return false;
            }
        }
        return true;
    };
    Ng2FileInputComponent.prototype.generateId = function () {
        return 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx'.replace(/[xy]/g, function (c) {
            var r = Math.random() * 16 | 0, v = c == 'x' ? r : (r & 0x3 | 0x8);
            return v.toString(16);
        });
    };
    Ng2FileInputComponent.decorators = [
        { type: core.Component, args: [{
                    selector: 'ng2-file-input',
                    template: "<div class=\"ng2-file-input\">\n                    <div class=\"ng2-file-input-invalid text-danger\" [hidden]=\"!invalidFile\" [innerHTML]=\"invalidFileText\"></div>\n                    <div fileDrop class=\"ng2-file-input-drop-container\" [ngClass]=\"{'file-is-over': fileIsOver}\" (fileOver)=\"fileOver($event)\"\n                        (onFileDrop)=\"onFileDrop($event)\">\n                        <span [innerHTML]=\"dropText\"></span>\n                        <button type=\"button\" (click)=\"ng2FileInputSelect.click()\" class=\"btn btn-primary\" [innerHTML]=\"browseText\"></button>\n                    </div>\n                    <div class=\"ng2-file-input-files\" *ngIf=\"showPreviews\">\n                        <div *ngFor=\"let file of getCurrentFiles()\" class=\"ng2-file-input-file\" [ngClass]=\"{'image':file.type.indexOf('image')!==-1}\">\n                            <span [innerHTML]=\"file.name\" class=\"ng2-file-input-file-text\"></span>\n                            <img [src]=\"getObjectUrl(file)\" *ngIf=\"file.type.indexOf('image')!==-1\">\n                            <span class=\"ng2-file-input-file-text remove\" (click)=\"removeFile(file)\" *ngIf=\"removable\">\n                                <p [innerHTML]=\"removeText\"></p>\n                            </span> \n                        </div>\n                    </div>\n                    <input type=\"file\" #ng2FileInputSelect (change)=\"fileSelected($event)\" [accept]=\"accept\" [attr.multiple]=\"multiple ? true : null\">\n                </div>",
                },] },
    ];
    /** @nocollapse */
    Ng2FileInputComponent.ctorParameters = function () { return [
        { type: FileInputHandlerService, decorators: [{ type: core.Inject, args: [core.forwardRef(function () { return FileInputHandlerService; }),] },] },
        { type: platformBrowser.DomSanitizer, decorators: [{ type: core.Inject, args: [core.forwardRef(function () { return platformBrowser.DomSanitizer; }),] },] },
        { type: Ng2FileInputOptions, decorators: [{ type: core.Inject, args: [core.forwardRef(function () { return Ng2FileInputOptions; }),] },] },
    ]; };
    Ng2FileInputComponent.propDecorators = {
        "id": [{ type: core.Input },],
        "dropText": [{ type: core.Input, args: ['drop-text',] },],
        "browseText": [{ type: core.Input, args: ['browse-text',] },],
        "removeText": [{ type: core.Input, args: ['remove-text',] },],
        "invalidFileText": [{ type: core.Input, args: ['invalid-file-text',] },],
        "invalidFileTimeout": [{ type: core.Input, args: ['invalid-file-timeout',] },],
        "multiple": [{ type: core.Input, args: ['multiple',] },],
        "accept": [{ type: core.Input, args: ['accept',] },],
        "removable": [{ type: core.Input, args: ['removable',] },],
        "showPreviews": [{ type: core.Input, args: ['show-previews',] },],
        "extensions": [{ type: core.Input, args: ['extensions',] },],
        "outputAction": [{ type: core.Output, args: ['onAction',] },],
        "outputRemoved": [{ type: core.Output, args: ['onRemoved',] },],
        "outputAdded": [{ type: core.Output, args: ['onAdded',] },],
        "outputInvalidDenied": [{ type: core.Output, args: ['onInvalidDenied',] },],
        "outputCouldNotRemove": [{ type: core.Output, args: ['onCouldNotRemove',] },],
        "outputCouldNotAdd": [{ type: core.Output, args: ['onCouldNotAdd',] },],
    };
    return Ng2FileInputComponent;
}());

var Ng2FileInputService = (function () {
    function Ng2FileInputService(fileInputHandlerService) {
        this.fileInputHandlerService = fileInputHandlerService;
    }
    Ng2FileInputService.prototype.reset = function (id) {
        return this.fileInputHandlerService.reset(id);
    };
    Ng2FileInputService.prototype.remove = function (id, files) {
        return this.fileInputHandlerService.removeFiles(id, files);
    };
    Ng2FileInputService.prototype.add = function (id, files) {
        return this.fileInputHandlerService.addFiles(id, files);
    };
    Ng2FileInputService.prototype.getCurrentFiles = function (id) {
        var fileInput = this.fileInputHandlerService.getFileInput(id);
        return fileInput ? fileInput.currentFiles : [];
    };
    Ng2FileInputService.decorators = [
        { type: core.Injectable },
    ];
    /** @nocollapse */
    Ng2FileInputService.ctorParameters = function () { return [
        { type: FileInputHandlerService, decorators: [{ type: core.Inject, args: [core.forwardRef(function () { return FileInputHandlerService; }),] },] },
    ]; };
    return Ng2FileInputService;
}());

var USER_OPTIONS = new core.InjectionToken('ng2 file input custom user options');
function optionsFactory(userOptions) {
    var options = new Ng2FileInputOptions();
    Object.assign(options, userOptions);
    return options;
}
var Ng2FileInputModule = (function () {
    function Ng2FileInputModule() {
    }
    Ng2FileInputModule.forRoot = function (options) {
        if (options === void 0) { options = {}; }
        return {
            ngModule: Ng2FileInputModule,
            providers: [{
                    provide: USER_OPTIONS,
                    useValue: options
                }, {
                    provide: Ng2FileInputOptions,
                    useFactory: optionsFactory,
                    deps: [USER_OPTIONS]
                }]
        };
    };
    Ng2FileInputModule.decorators = [
        { type: core.NgModule, args: [{
                    imports: [
                        common.CommonModule,
                        bergbenAngular2FileDrop.FileDropModule
                    ],
                    providers: [Ng2FileInputService, FileInputHandlerService],
                    declarations: [Ng2FileInputComponent],
                    exports: [Ng2FileInputComponent]
                },] },
    ];
    /** @nocollapse */
    Ng2FileInputModule.ctorParameters = function () { return []; };
    return Ng2FileInputModule;
}());

exports.Ng2FileInputComponent = Ng2FileInputComponent;
exports.FileInputHandlerService = FileInputHandlerService;
exports.Ng2FileInputService = Ng2FileInputService;
exports.Ng2FileInputOptions = Ng2FileInputOptions;
exports.Ng2FileInputModule = Ng2FileInputModule;

Object.defineProperty(exports, '__esModule', { value: true });

})));
